design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,323 @@
+succeeded in this timeout, the request fails. The next request `B` receives, it will again try to 
+forward it to `A`, since `A` is still the active, and it will fail again. This will happen until 
+rebalancing completes and a new active is elected for the partition. 
+
+There are two steps in the rebalancing procedure that are expensive: 
+  1) One of `B` or `C` is picked as the new active, this takes ~10 seconds based on default configs. 
+  2) Once a new active is chosen, depending on how far behind (lag) it is with respect to the changelog, 
+  it may take up  few seconds or even minutes before the standby has fully caught up to the last committed offset. 
+
+In total, it takes >>10 seconds before a query can start succeeding again. What can we do until a
+new active is ready to serve requests? 
+[KIP-535](https://cwiki.apache.org/confluence/display/KAFKA/KIP-535%3A+Allow+state+stores+to+serve+stale+reads+during+rebalance) 
+has laid the groundwork to allow us to serve requests from standbys, even if they are still in 
+rebalancing state or have not caught up to the last committed offset of the active. 
+
+Every KSQL server now has the information of:
+
+1. Local offset lag: The lag of all partitions the server hosts (whether active or standby)
+2. Routing Metadata for a key: The partition, active host and standy hosts per key
+
+This information allows each server (with some communication, discussed below) to know the global 
+lag information of every topic partition whether it is the standby or active. This enables us to 
+implement a policy where `B` having established that `A` is down (after trying to send a request 
+to `A` and seeing it failed), to decide whether it (`B`) will serve it or `C`. This effectively 
+means that there will be little to no down time in serving requests at the cost of consistency as
+`B` or `C` may serve stale (not caught up) data. Hence, we achieve high availability at the cost of 
+consistency. Eventual consistency for the win!
+
+
+## What is in scope
+Ensure availability for pull queries when KSQL servers fail (with `ksql.streams.num.standby.replicas=N`, 
+we can tolerate upto N such failures) 
+
+## What is not in scope
+
+- Address unavailability caused by cold starts of KSQL server i.e when a new server is added to a 
+KSQL cluster, it must first rebuild its state off the changelog topics and that process still could 
+take a long time. 
+- Try to improve consistency guarantees provided by KSQL i.e reduce the amount of time it takes to 
+rebalance or standby replication to catch up. 
+
+## Value/Return
+
+The cluster of KSQl server will be able to serve requests even when the active is down. This mitigates 
+a large current gap in deploying KSQL pull queries for mission critical use-cases, by significantly 
+reducing failure rate of pull queries during server failures.
+
+## Public APIS
+No change to public APIs
```

Comment:
Should users not have a configuration that allow them to define "how stall" data may be? Adding new configs would be a public API change?

Reply:
In a server model, a server config won't be an API change. However, I think there was feedback to make this a per-table config, at which point this needs to called out.. @vpapavas ? wdyt

Reply:
I think we'd normally call out server config in this section.

As discussed, I think allowing the client doing the pull query to specify the acceptable lag, (or min event time), would be more powerful that a one-size-fits-all server config, even if that config is per-table.  Consider a table where many different use-cases query it.  Some may be OK with stale data, others not.   Such a per-call config could override any server side setting.

Reply:
Hey @big-andy-coates , the text of the KLIP is updated to include the per query configuration parameter.